drivers: ieee802154: introduce channel pages

fgrandel opened the pull request:
# TLDR;

This PR fixes some non-standard conforming aspects of the IEEE 802.15.4 driver API by introducing the standard conforming concept of _channel pages_ (see IEEE 802.15.4-2020, section 10.1.3).

~~*Obs: The DNM label was only attached to ensure that all driver maintainers have a fair chance to review this before it's being merged. No reason not to review this PR.*~~

UPDATE: Removed DNM tag as all important reviews have come in. @jciupis As you're on vacation now, @rlubos was so kind to take over. :-)

Driver maintainers: Ideally you try this out on real hardware before ACK. I already did so for the CC13/26xx SubG driver myself.

# Detailed Rationale

This is the lowest hanging fruit in the direction of RFC #61227 (vendor/L2 agnostic driver API) which was recently agreed upon in the Arch WG. It is a good example of how the principles introduced there immediately make the driver API more vendor-agnostic.

Introducing channel pages requires non-trivial changes to the driver API and must therefore be well justified:
 * The previous approach did not support all existing drivers, namely those working on channel pages 2/5 (780/863/915 MHz O-QPSK - rf2xx), 9 (SUN FSK - cc120x and cc13/26xx_subg) and 4 (HRP UWB - dw1000). Enabling SUN FSK and UWB support alone justifies this change IMO:
   * HRP UWB is becoming increasingly important as it enables precision ranging (see the work of UWB alliance and FiRa re IEEE 802.15.4-2020z).
   * SUN FSK is common in the industrial IoT space and increasingly replaces O-QPSK/BPSK in the SubG bands (see the WiSUN standard built on top of it).
 * The new approach is required to correctly determine several PHY dependent timing parameters. It fixes CSMA/CA timing errors for existing drivers on channel pages 2, 4, 5 and 9.
 * The new approach removes a long-standing restriction in scanning for networks. Previously only 2.4G networks on channel page zero could be scanned - drivers supporting other channels or channel pages were not supported. Now scanning is generically supported for all existing and potential future drivers.
 * The new approach allows driver maintainers to specify multiple and/or non-zero based channel ranges per driver and/or channel page. This is immediately useful for the existing DW1000, SubG and O-QPSK 2.4G drivers but also relevant in the SUN FSK area in the future where not all channel pages have zero-based channel ranges.
 * The new approach removes the necessity to write multiple drivers per chipset if that chipset supports several bands (e.g. the separation between the cc13xx_cc26xx and cc13xx_cc26xx_subg drivers could now be removed).
 * Channel pages are required to support channel hopping which is relevant for several sub-protocols of the IEEE 802.15.4 standard, namely DSME, TSCH and SUN (but also generically supported by other sub-protocols).
 * Channel pages are required to support Spectrum Resource Measurement (SRM), i.e. measuring network resource usage and performance.
 * The change ensures that drivers for transceivers implementing completely new PHYs (LRP UWB, MSK, TVWS, ...) will be supported ootb as we now implement standard primitives that _all_ existing and future PHYs must comply with. The API is thereby made more future proof.

This change makes use of the recently introduced PHY PIB attributes getter and hopefully demonstrates and proves its usefulness.

Commit:
drivers: ieee802154: introduce channel pages

Replaces the previous approach to define bands via hardware capabilities
by the standard conforming concept of channel pages.

In the short term this allows us to correctly calculate the PHY specific
symbol rate and several parameters that directly depend from the symbol
rate and were previously not being correctly calculated for some of the
drivers whose channel pages could not be represented previously:
* We now support sub-nanosecond precision symbol rates for UWB. Rounding
  errors are being minimized by switching from a divide-then-multiply
  approach to a multiply-then-divide approach.
* UWB HRP: symbol rate depends on channel page specific preamble symbol
  rate which again requires the pulse repetition value to be known
* Several MAC timings are being corrected based on the now correctly
  calculated symbol rates, namely aTurnaroundTime, aUnitBackoffPeriod,
  aBaseSuperframeDuration.

In the long term, this change unlocks such highly promising functional
areas as UWB ranging and SUN-PHY channel hopping in the SubG area (plus
of course any other PHY specific feature).

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

cfriedt reviewed (commented):
Looks overall agreeable.

Great work with the cc13xx and enumerating pages.

Maybe not needed in this rev, and perhaps it's already available, but it would be good to have DT attributes for the radio device to allow board-Level overrides (there are almost always board level overrides).

I'll save my approval for after others have had a round.

kartben commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -30,101 +30,319 @@ extern "C" {
+ * Default macResponseWaitTime in multiples of aBaseSuperframeDuration as
+ * defined in section 8.4.3.1, table 8-94.
+ */
+#define IEEE802154_MAC_RESONSE_WAIT_TIME_DEFAULT 32U
```
```suggestion
#define IEEE802154_MAC_RESPONSE_WAIT_TIME_DEFAULT 32U
```

fgrandel replied:
Fixed, thanks for spotting this.

fgrandel commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:
```diff
@@ -514,6 +508,18 @@ ieee802154_cc13xx_cc26xx_configure(const struct device *dev,
 	return -ENOTSUP;
 }
 
+/* driver-allocated attribute memory - constant across all driver instances */
+IEEE802154_DEFINE_PHY_SUPPORTED_CHANNELS(attributes, 11, 26);
+
+static int ieee802154_cc13xx_cc26xx_attr_get(const struct device *dev, enum ieee802154_attr attr,
+					     struct ieee802154_attr_value *value)
+{
+	ARG_UNUSED(dev);
+
+	return ieee802154_attr_get_channel_page_and_range(
+		attr, IEEE802154_ATTR_PHY_CHANNEL_PAGE_ZERO_OQPSK_2450_BPSK_868_915,
+		&attributes.phy_supported_channels, value);
+}
```
@cfriedt Agreed. DT-level overrides are an important point but only part of the picture IMO. This PR only scratches the surface of the rich universe of PHY configuration options (standard and custom). Guess you were also not so much stressing DT but more that we need enough degrees of freedom to cater for vendor- and project-specific extensions. If that's what you mean - also 100% agreed.

Right now we just have much less need for non-standard/vendor specific extensions in existing drivers than it seems. Most of the existing configuration options are standard options in the wrong place. This PR is proof of it but there's a lot more. Of course vendor/protocol specific options on all levels also need to be catered for (see 2, 5, 6 below) even if it's the smaller problem IMO, as those can easily be encapsulated and we already have a good framework for it (since a few weeks).

What follows is mostly "note to self" - no need to read the details. I just like the opportunity to jot down my current thoughts in public so people can comment and I can refer to it in the future and let my bad memory forget it for now.

The PHY specification is more than half of the standard and is still the fastest growing part. First we need to thoroughly understand the structure of the IEEE 802.15.4 PHY configuration space and then we have to systematically map it onto existing Zephyr architecture elements. Once we have the full picture it will be easy to find the "right" place for each and every requirement that might come up in the future - it may be DT it also may be not. Luckily the standardization committee has done most of the hard work, [we just need to apply their concepts](https://github.com/zephyrproject-rtos/zephyr/issues/61227).

Here comes my current mind model of the required multi-dimensional IEEE 802.15.4 PHY configuration space which embeds DT but goes beyond. It is spanned by several orthogonal dimensions. Please let me know what I'm missing:

1. Granularity
    * Global defaults: 100% standard based (this is mostly what this PR starts off)
    * Global (const): Kconfig
    * Per driver (const): Kconfig
    * Per driver instance/interface (const): DT
    * Per driver instance/interface (dynamic): net_mgmt & driver_api->configure()

2. Horizontal Layers
    * PHY aspects of driver hardware
    * PHY aspects of driver firmware (may be multiple per hardware)
    * L2 <-> PHY configuration interfaces
    * upper layers <-> PHY configuration interfaces (to be avoided - but bad design of upper layers may force us)

3. Vertical Stacks

    There are very few differences between native L2 and Thread relating to PHY configuration. Thread only fills a few gaps in the standard to improve cross-vendor compatibility. Standard protocols (vanilla WPANs, DTME, TSCH, CSL, RIT, ranging, SRM, etc.) are decoupled from PHY by design, so probably no need for extra PHY config there. We need to cater for a few OT implementation quirks unrelated to the Thread standard and we already have an architecture to do so. The same mechanism will work for PHY extensions/refinements of protocols we may want to support in the future (ISA 100.11a, ...).

4. PHY

    This PR only gives a glimpse of what can be configured on a per PHY level. Some of the PHYs are mostly standardized (e.g. O-QPSK, BPSK), others are highly configurable (SUN FSK, LECIM, TVWS, to a lesser extent UWB). Providing a good configuration model requires intimate knowledge of the PHY specific physics, modulation and coding approach which is challenging. The best we can do for now is ensuring that the configuration architecture includes well encapsulated driver-agnostic "configuration namespaces/containers" for each PHY. Luckily the standard has done the hard work of extracting what is common across PHYs and pushing down and prestructuring the PHY-specific data model. We just need to read and follow the standard here - this PR is an example of that approach.

5. Chipset

    Recent hardware is increasingly standard conforming. But there are some quirks. Vendors try to differentiate via non-standard options or try to create their own de-facto standard. TI's CCxxxx (from 11xx up to the latest 26x7) is a good example as it is highly configurable on PHY level which makes it very flexible for custom applications. All datasheets I've opened document at least a few non-standard options. The configuration can easily be isolated in this case in driver specific defaults, Kconfig, DT and our new API for well-encapsulated custom extensions to runtime configuration. So this is not a problem in practice. For non-standard options I propose we be tolerant and live the diversity to resist premature abstractions. If patterns emerge in practice we can still iterate upon them to make them driver agnostic.

    The opposite is true for standard options: We need to be much more careful in the future to extract configuration defined by the standard that is per definition NOT hardware specific such that drivers don't re-invent the wheel all the time (as can be seen in current drivers). We can repeat the same O-QSPK symbol rate in 10 defines in the code - I don't mind. But exposing users to 10 configuration options - one per driver - that essentially all manipulate the same standard PHY PIB attribute is really bad.

6. Firmware

    Vendor specific driver firmware can have tons of configuration options - some standard some not (nRF5 is a good example). Our challenge here will again be to extract what is part of the standard so that users wanting to switch between drivers or using multiple of them don't have to learn each and every driver firmware configuration model. Of course there are limits to what makes sense and we should resist the temptation for premature optimization. Again my approach would be to live diversity and wait for patterns to emerge when it comes to non-standard configuration options. But we should be very strict in the future to extract standard options whenever possible rather than letting driver maintainers simply wrap their vendor-specific and mostly non-reusable custom API for those.

Of course none of this should be introduced just for the sake of it but strictly on an as-needed basis. I'm spanning up this framework to avoid the mistake of the past that the first driver requiring some configuration inside this space establishes it in some driver specific Kconfig variable which we then have to support forever although it was probably some standardized PHY configuration that would have much better been supported via interface-specific DT or driver agnostic runtime configuration.

**My intention is all about the user facing API becoming simpler, more vendor agnostic, future proof and stable.** That's also why I'm being a little defensive, conservative and slow - we've been shooting a little to quickly in the past which now causes us some extra work and headache (see this PR) - but that's also normal and not a big problem IMO.

jciupis commented on subsys/net/l2/ieee802154/ieee802154_utils.h:
```diff
@@ -278,112 +278,204 @@ static inline void ieee802154_radio_remove_pan_id(struct net_if *iface, uint16_t
+ * not configured by L2 but directly provided by the driver.
  *
- * @param iface The interface for which the symbol period should be calculated.
+ * @param iface The interface for which the turnaround time should be calculated.
```
You probably meant channel page here.

fgrandel replied:
Fixed.

jciupis commented on subsys/net/l2/ieee802154/ieee802154_utils.h:
```diff
@@ -278,112 +278,204 @@ static inline void ieee802154_radio_remove_pan_id(struct net_if *iface, uint16_t
+
+	/* Section 11.3, table 11-1, aTurnaroundTime: "For the SUN [...] PHYs,
+	 * the value is 1 ms expressed in symbol periods, rounded up to the next
+	 * integer number of sym- bol periods using the ceiling() function. [...]
```
`symbol`

fgrandel replied:
Fixed.

jciupis commented on subsys/net/l2/ieee802154/ieee802154_utils.h:
```diff
@@ -278,112 +278,204 @@ static inline void ieee802154_radio_remove_pan_id(struct net_if *iface, uint16_t
+ * @brief Counts all available channels of the driver's currently selected
+ * channel page.
+ *
+ * @param iface The interface for which the turnaround time should be calculated.
```
Please reword the documentation of this parameter here and in other functions (e.g. `ieee802154_radio_verify_channel`) where turnaround time is referred to out of context.

fgrandel replied:
Fixed.

jciupis reviewed (commented):
The PR looks good overall, I only spotted some minor typos and mistakes in the doc. I'm not approving yet to verify on hardware first, as requested.

fgrandel commented on drivers/ieee802154/ieee802154_dw1000.c:
```diff
@@ -940,6 +947,46 @@ static int dwt_configure(const struct device *dev,
+		.ranges = attributes.phy_supported_channel_ranges,
+		.num_ranges = 2U,
+	},
+};
```
Hi @jfischer-no , thanks for your review. :-)

Sure, I can change those names to shorter ones, although the naming topic is much more involved than it may seem.

Those names have been chosen carefully in line with #61227 - so if we shorten them, we still need to work by these rules:
* Inside the driver distinction between PHY vs. MAC is mandatory unless the context is clear. Attributes can be defined in both contexts. Therefore "phy_attr" would be wrong.
* Names need to refer to IEEE 802.15.4 standard primitives if applicable. This is the case for the "channel" and "channel page" terms.
* Names must designate non-ambivalent points in the standard's global configuration space even if we only implement it partially for now. This PR was only necessary because we didn't follow that rule in the first place! This means that we need to give additional context to the "channel" term.

Additionally: Using abbreviations is not a mandatory coding guideline in Zephyr AFAICS. I personally consider them a maintainability problem on most occasions when I encounter them in Zephyr. After all we optimize maintainability (i.e. TCO) and not variable name length - the rest is up to proper IDE settings.

All that said, I do agree, though, that you've a point. My naming is inconsistent - thanks for pointing me to that.

I therefore changed

```
static const struct {
	const struct ieee802154_phy_channel_range phy_supported_channel_ranges[2];
	const struct ieee802154_phy_supported_channels phy_supported_channels;
} attributes ...
```

to

```
static const struct {
	const struct ieee802154_phy_channel_range phy_channel_range[2];
	const struct ieee802154_phy_supported_channels phy_supported_channels;
} attributes ...
```

(here and in the default DEFINE_... template)

Is there anything more you think we can still shorten, given the above information?

jfischer-no replied:
TL;DR; Members of `static const struct` can be just `r` and `c`.  What matters is the name of parent `static const struct`.  `attributes` is too common in the context of a driver and should better be prefixed. `attr` is a widely used abbreviation in the tree.

fgrandel replied:
@jfischer-no As you insist: Prefixed and renamed _attributes_ to _drv_attr_.  For the rest you please first read then comment. TLDR not accepted.

jfischer-no replied:
`drv_attr` is still not a correct name as is probably misleading, how I understood your patch,  struct ieee802154_phy_channel_range and struct ieee802154_phy_supported_channels are to describe phy/radio-controller capabilities and not drivers limitations. So this should be phy_attr or phy_attribute, and there is no reason to mix mac attributes in the same `static const structure` or how it could be not distinct where everything related is prefix with `ieee802154_phy_.`

fgrandel replied:
You seem not to be aware of #61227 which gives the context and explanation why this is correct (driver = PHY + hard MAC). Dismissed as wrong from a maintainer's perspective.

jfischer-no replied:
Okay. I try to read it a few times, but TBH I cannot find this outcome. Further, the statement `driver = PHY + hard MAC` is very problematic. There is no driver for the HardMAC device in drivers/ieee802154, no matter what feature a radio controller has. I was under the assumption that it was cleaned up and layering improved, but I see it staying as it is.

fgrandel replied:
> There is no driver for the HardMAC device

The driver with most hard MAC features and the group from which the most concern re hardware offloading is actually (rightfully) coming, is Nordic's. As you seem to be working for them, I propose you co-ordinate yourself internally with @rlubos on that topic. He might be able to explain the Nordic requirements better to you. He's also been in the loop of our discussions all the time and should be able to explain the architectural direction to you as well as I can.

fgrandel commented:
@jciupis @nandojve @jfischer-no @cfriedt Could you (re-)review please? Hope this can be finalized now?

rlubos reviewed (dismissed):
LGTM

kartben commented on include/zephyr/net/ieee802154_radio.h:
```diff
@@ -31,72 +31,311 @@ extern "C" {
- *       https://github.com/zephyrproject-rtos/zephyr/issues/50336#issuecomment-1251122582.
- *       For now we assume PHYs that current drivers actually implement.
+/**
+ * MAC services (section 8)
```
it looks as if you would either want to have this not be a doxygen comment, or instead create `@name`d sections, since I don't think you want "MAC services (section 8)" to only show up for `IEEE802154_MAC_A_BASE_SLOT_DURATION`. This comment applies to basically all the other "(section XXX)" comments.

<img width="773" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/322d071d-9aa0-4507-a8c7-96d69acd15e8">

fgrandel replied:
Oh, and if you find similar issues in #62503 then of course it would be great to point me to it as that one really intends to fix all of them as far as possible.

kartben replied:
> @kartben Is it ok if we postpone all documentation issues until this one is merged, as it will immediately unblock #62503 which fixes this problem and lots and lots of other documentation issues as well?

Absolutely :)

fgrandel commented:
> Hi @fgrandel ,
> 
> I saw that you removed the DNM label. When we talked a few days ago I mentioned that I can only start to check on HW this weekend because I was on a trip due to a summit. It is a surprise to me that now this PR start to be in rush since we agreed that it is more important test on HW and make sure that everything is fine than rush and start to broke things.
> 
> What changed?

Wasn't there a weekend in between since we talked - misunderstanding, sorry. :-) Apart from that: nothing changed. I'll put the DNM tag back and wait until next week.

nandojve commented:
> > I saw that you removed the DNM label. When we talked a few days ago I mentioned that I can only start to check on HW this weekend
> 
> I thought you were talking about last weekend. Just a misunderstanding. I'll put the DNM tag back and wait until next week.

Maybe it was confusing because we first talked at 08/09 when I was in another trip and than we sync once more during the week. After that I travel again to the summit and didn't have time to touch HW. Last weeks have been quite busy for me. Thank you for understand it.

nandojve commented:
Hi @maxmclau ,

I think you may be interested on this changes.

cfriedt commented:
@vaishnavachath also impacts beagle things (in a good way, I hope).

Although it isn't clear to me if the Linux kernel has support yet

jfischer-no commented on subsys/net/l2/ieee802154/ieee802154_utils.c:
```diff
@@ -0,0 +1,72 @@
+		return false;
+	}
+
+	for (int channel_range_index = 0;
+	     channel_range_index < value.phy_supported_channels->num_ranges;
+	     channel_range_index++) {
+		const struct ieee802154_phy_channel_range *const channel_range =
+			&value.phy_supported_channels->ranges[channel_range_index];
+
+		if (channel >= channel_range->from_channel &&
+		    channel <= channel_range->to_channel) {
+			return true;
+		}
+	}
```
This can be just `	for (int idx = 0; idx < value.phy_supported_channels->num_ranges; idx++) {`
and before you complain this is just my personal attitude (which it certainly is, and I have learned from contributing a few thousand lines of code to 15.4 drivers from the maintainers), please [RTFCS](https://www.kernel.org/doc/html/latest/process/coding-style.html#naming).

jfischer-no replied:
If you were to read our coding style, you would find that "In general, follow the [Linux kernel coding style](https://kernel.org/doc/html/latest/process/coding-style.html), with the following exceptions:..." https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

Further, I wrote "This can be just...", not "This must be...". IMHO, spreading information unnecessarily over a large number of characters is not good practice, either in coding or in writing RFCs or responses.

fgrandel replied:
Thanks for the pointer - very useful.

jfischer-no commented on subsys/net/l2/ieee802154/ieee802154_utils.c:
```diff
@@ -0,0 +1,72 @@
+		return 0;
+	}
+
+	for (int channel_range_index = 0;
```
`for (int idx = 0; idx < value.phy_supported_channels->num_ranges; idx++) {`

jfischer-no reviewed (changes requested):
This patch has some stray and not directly related to the commit message changes,  and needs some additional work.

Commit:
drivers: ieee802154: cc1200: SUN-FSK compatibility

Preparative change to introduce build-time configured channel pages.
This fixes the description of the driver's available PHYs and makes
channel page and channel range independent from runtime attributes.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: cc13xx_cc26xx: readability improvement

Aligns the name of the return value variable with what is used elsewhere
in the driver and the subsystem for improved readability and
consistency.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@nandojve @rlubos Would you mind re-approving? I had to push again due to some valid feedback from @jfischer-no 's second review. Nothing functional at all just some shuffling around of commits and a better commit message. I made sure that you can confirm this by clicking on the compare button above so that you do not unnecessarily waste your time.

vaishnavachath commented:
> @vaishnavachath , @jadonk - also impacts beagle things (in a good way, I hope).
> 
> Although it isn't clear to me if the Linux kernel has support yet

@cfriedt @fgrandel , Thanks the changes looks good, yes this helps the wpanusb/serial support in a good way. we had some hack/Kconfigs to do few of these things : https://git.beagleboard.org/beagleconnect/zephyr/wpanusb_bc/-/blob/master/prj.conf?ref_type=heads#L32 and this set of changes helps to make it much cleaner.

nandojve commented:
Hi @fgrandel ,

Organizational topic idea: Is possible to create a GH project or an issue that allows us to track current/future enhancements? My intention is not to read the content/feature itself but instead have the bullets that we know what will be in one or another release, for instance. The IEEE 802.15.4 seems to growing these days and it will be nice understand what Feature/PR goes on one release to ensure proper HW tests. This could help us to look and understand what it is implemented and not too.

In this https://github.com/zephyrproject-rtos/zephyr/issues/38657 there is a list of features that it is easy to track.

Maybe we need both issue to track IEEE 802.15.4 features and a project to understand what goes in the releases.

Feel free to implement if make sense : )

fgrandel commented:
> Organizational topic idea: Is possible to create a GH project or an issue that allows us to track current/future enhancements?

@nandojve If I understand you correctly, you're proposing that we maintain a proper subsystem roadmap, right? This is a great idea - thanks for that proposal. And I love it even more that we conceive this roadmap as an open community effort that we jointly maintain as you propose.

I'm currently preparing a [presentation for the Linux Plumbers Conference](https://docs.google.com/presentation/d/1o5HDS8cWxj4zYPC91tbLxiWQTHZjo8nxi18OiDyJhBY ) in November where I collect all changes that happened over the last months (by topic area) plus I'm presenting the opportunities that I see in the subsystem for the upcoming releases that have been unlocked by these past changes. Be aware that this is early work in progress but feel free to comment.

> Maybe we need both issue to track IEEE 802.15.4 features and a project to understand what goes in the releases.

Some things already exist:
 * I already do have a [public project for the TSCH effort](https://github.com/orgs/zephyr-tsch/projects/3/views/1). But I agree that it makes sense to **transform this into a IEEE 802.15.4 subsystem roadmap as soon as other contributors take on long-running projects**.
 * Some of the topics on this roadmap have been refined via RFC process:
   * Fundamental architectural direction of the subsystem: #61227
   * TSCH: #50336
   * Abstract high resolution clocks (required for all kinds of precision timing, including TDMA protocols like TSCH): https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731
   * Drill down into the syntonization layer of that proposed subsystem ([reference implementation](https://github.com/fgrandel/zephyr/commit/af20bed42c7a3bad9a5c3a93576d6cf572b10edf) already available [for the CC13/26xx SubG driver](https://github.com/fgrandel/zephyr/commit/5b965aa4cfb5460da318fbdc02fff99e6dbaa4ab) in my dev branch): #60400

**Do you personally plan any project for the future that I should add to that roadmap?** If that's the case I'll do so immediately.

Some more background for those who stumble upon this comment: @nandojve and I identified the following **potential opportunities for driver maintainers** (this list is tailored to the rf2xx driver and of course far from complete, just to give you the rough idea):
* **OpenThread support**: Test the OpenThread L2 with additional drivers, identify issues and find solutions (unlocked by ongoing changes that simplify and standardize the driver API for cross-vendor support, see #61227). Robert as the OT maintainer already offered to help any driver maintainer being ambitious in that area.
* **Additional PHY or driver features**, namely:
   * support zero-copy communication between L2 and the driver by letting the driver announce head-/tailroom requirements (unlocked by the new attributes infrastructure).
   * support larger frames offered by PHYs as SUN or MSK with faster bit rates (the standard allows up to 2047 bytes IIRC, jointly unlocked by the new driver attributes feature and this PR)
* **Tooling**: Create IEEE 802.15.4 debugging tools, e.g. test existing in-tree or out-of-tree sniffing solutions and document them.

Apart form that ranging is a big opportunity in the subsystem - I'm inviting everyone in the community to look at FiRa/802.15.4z and extend the existing DW1000 driver in that direction or introduce new drivers (e.g. NXP) that support ranging.

jfischer-no commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx.c:
```diff
@@ -155,42 +155,36 @@ static inline int ieee802154_cc13xx_cc26xx_channel_to_frequency(
+	/* See IEEE 802.15.4-2020, section 10.1.3.3. */
+	if (channel >= 11 && channel <= 26) {
+		*frequency = 2405 + 5 * (channel - 11);
 		*fractFreq = 0;
+		return 0;
 	} else {
+		/* TODO: Support sub-GHz for CC13xx rather than having separate drivers */
 		*frequency = 0;
 		*fractFreq = 0;
-		return -EINVAL;
+		return channel < 11 ? -ENOTSUP : -EINVAL;
 	}
-
-	return 0;
```
Looks like these changes are not covered by the commit message and probably belong in the next commit, (similar changes https://github.com/zephyrproject-rtos/zephyr/pull/62413/commits/c5684a43c04945d19d5dedbac099e402a3b6e194#diff-8278a3f5641432256702d96822ad636a501df47e873791c58ccdd79b77cad1bc). Cleaner would be to move the removal of `IEEE802154_*_CHANNEL_MAX`/`IEEE802154_*_CHANNEL_MIN` to own commit as they are only used in the _cc13xx_cc26xx_ drivers.

jfischer-no replied:
@fgrandel Stop to mark resolved unresolved conversations. You are not doing your patches and commits properly. This is not how patches submitted by a maintainer should look. I am clearly showing you a problem in your PR that is still unresolved. And no, you cannot ignore them, because you are the maintainer of that area. These are basic things you learn as a contributor and collaborator: to break your work into smaller, self-contained changes, to have pre-work patches before major changes, and to format and describe them in a way that can be followed just by looking at the git diffs and logs.

nashif replied:
@fgrandel this request seems to be valid, please just move this hunk to the next commit and let's move on with this.

@jfischer-no Please slow down and take a breath. This is the longest comment I have ever seen about a misplaced hunk in a commit.

jfischer-no replied:
> @jfischer-no Please slow down and take a breath. This is the longest comment I have ever seen about a misplaced hunk in a commit.

I did not even speed up to slow down. My comment about the inability to submit a reasonable set of patches, compared to the long novellas without significant information content that you otherwise find here, is very concise.

fgrandel replied:
@nashif 

> this request seems to be valid, please just move this hunk to the next commit and let's move on with this.

True. Fixed.

fgrandel commented:
> This patch has some stray and not directly related to the commit message changes, and needs some additional work.

@jfischer-no As we seem to agree by now, that you've been providing non-blocking feedback only and I've already made my choice of what I'd like to take over and what not as you can see. I therefore kindly ask you to unblock the PR ASAP as it contains an important contribution that all relevant reviewers from the subsystem have approved and that blocks several dependent PRs.

I've taken note of the relevance of the Linux Coding Guidelines in Zephyr so I hope this topic has now found due discussion in this PR which is otherwise unrelated.

cfriedt commented:
> Organizational topic idea: Is possible to create a GH project or an issue that allows us to track current/future enhancements? My intention is not to read the content/feature itself but instead have the bullets that we know what will be in one or another release, for instance. The IEEE 802.15.4 seems to growing these days and it will be nice understand what Feature/PR goes on one release to ensure proper HW tests. This could help us to look and understand what it is implemented and not too.

@nandojve, @fgrandel - you might want to simply use the release plan for feature tracking. That would make the most sense and at least would keep release managers in the loop implicitly.

https://github.com/orgs/zephyrproject-rtos/projects/13/views/1

nashif commented:
@jfischer-no  please revisit

jfischer-no commented on doc/releases/release-notes-3.5.rst:
```diff
@@ -177,6 +177,23 @@ Drivers and Sensors
 
 * IEEE 802.15.4
 
+  * A new mandatory method attr_get() was introduced into ieee802154_radio_api.
+    Drivers need to implement at least
+    IEEE802154_ATTR_PHY_SUPPORTED_CHANNEL_PAGES and
+    IEEE802154_ATTR_PHY_SUPPORTED_CHANNEL_RANGES.
+  * The hardware capabilities IEEE802154_HW_2_4_GHZ and IEEE802154_HW_SUB_GHZ
+    were removed as they were not aligned with the standard and some already
+    existing drivers couldn't properly express their channel page and channel
+    range (notably SUN FSK and HRP UWB drivers). The capabilities were replaced
+    by the standard conforming new driver attribute
+    IEEE802154_ATTR_PHY_SUPPORTED_CHANNEL_PAGES that fits all in-tree drivers.
+  * The method get_subg_channel_count() was removed from ieee802154_radio_api.
+    This method could not properly express the channel range of existing drivers
+    (notably SUN FSK drivers that implement channel pages > 0 and may not have
+    zero-based channel ranges or UWB drivers that could not be represented at
+    all). The method was replaced by the new driver attribute
+    IEEE802154_ATTR_PHY_SUPPORTED_CHANNEL_RANGES that fits all in-tree drivers.
+
```
It looks to me like a stable API Change (IEEE 802.15.4 driver interface API), should not part of it be added to  “API Changes” section of the release notes? To be fear there is not “API Changes” section in `doc/releases/release-notes-3.4.rst`.

I also could not find an RFC mail sent to the devel list about this, is there a new undocumented process or is it not needed here?

carlescufi replied:
That's a good point. However, I can't find the ieee802154 marked as stable here:
https://docs.zephyrproject.org/latest/develop/api/overview.html

>should not part of it be added to “API Changes” section of the release notes?

Stable API changes now belong in the migration guide, not in the release notes (hence why no section).

@fgrandel should we consider this API as unstable for now?

fgrandel replied:
@carlescufi Yes, one of the PRs blocked by this PR does just that. :-)

See #62503

The documentation has been less than satisfactory so far. And that PR fixes a lot of it.

fgrandel replied:
> This is part of networking APIs, stable since 1.0

I personally don't agree. The subsystem exposes three independent APIs: L2, driver and net_mgmt (each with their corresponding configuration approach in Kconfig, DT and runtime). As these three APIs have very different rythms of change and clients/exposure in applications I propose we treat those as separate.

@rlubos, @jukkar  and I considered the existing user base of those APIs small and the changes required for its transformation many so that we agreed to treat it de facto as unstable over the last three releases (which I'm proposing to formalize now).

This strategy seems to have worked out well. AFAICS we didn't get a single complaint and a lot of positive feedback for that approach from the community.

kartben replied:
> This is part of networking APIs, stable since 1.0 https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview [Networking APIs](https://docs.zephyrproject.org/latest/connectivity/networking/api/index.html#networking-api)

Unfortunately I don't think it's all that clear, and we don't really have a single source of truth for what's the status of each API within the Networking area.
For example that page you're linking also has CoAP, or MQTT, that are clearly labelled unstable, and many other entries that just don't happen to have ever been properly qualified in the documentation (ex. TFTP seems to be experimental if one looks at the Kconfig, but what should one make of it being listed on that page with no clear warning it's experimental?)

carlescufi replied:
I agree with the statements from @fgrandel and @kartben, but I will ask @rlubos and @jukkar to weigh in.

nandojve replied:
Hi @carlescufi ,

Just a remind that this IEEE 802.15.4 don't follow 100% the standard and many issues have been addressed recently. I would say the API is far to be stable and there are many changes in the backlog (which is not visible yet). My attempt to give the idea about a central point that have the features already in/planned goes in the direction to create some visibility about work already done and planned. As an example, the #49775 is already an example about things that we have been discussing over time and that already require possible changes/extensions in the API.

jfischer-no replied:
okay, it convinces me that it is okay to cherry-pick Radio API to be unstable.
But then why cannot [this change](https://github.com/zephyrproject-rtos/zephyr/pull/62503/files#diff-925c8ea16c15775440380c5e134bba4f8ba66aed774012b72b149e1b2210b884) be a commit before breaking changes, for the avoidance of any doubts?

rlubos replied:
I agree, the list of different components constituting "Networking API" is just too broad to have it covered by a single entry.

fgrandel commented:
> you might want to simply use the release plan for feature tracking

@cfriedt Good idea! Thanks for the pointer. I'll have a look at that project and see how I can interface existing activities with it.
